### DISCUSSION TITLE
Is there a way to use AVA with TypeScript + ESM without compile it?

### DISCUSSION THREAD

**Discussion by wellwelwel** [avajs/ava#3308] [Q&A] [answered]
I'm trying to perform a simple error test:

```bash
npm i -D ava tsx
```

> test/index.test.ts
```ts
import test from "ava";

test("Number should not be a text", (t) => {
  t.deepEqual("1", 1);
});
```

> package.json
```json
{
  "type": "module",
  "ava": {
    "files": [
      "test/**/*.test.ts"
    ],
    "extensions": {
      "ts": "module"
    },
    "nodeArguments": [
      "--import=tsx"
    ]
  },
  "devDependencies": {
    "ava": "^6.1.1",
    "tsx": "^4.7.1"
  }
}
```

> tsconfig.json
```json
{
  "compilerOptions": {
    "target": "ESNext",
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
    "esModuleInterop": true
  }
}
```

---

The above test works if I change from:
```diff
- "ts": "module"
```

To:
```diff
+ "ts": "commonjs"
```

But, naturally, running it as **CJS** instead.

---

❌ Here's the error when running it as `"module"`:
```bash
npx ava
```

```
  Uncaught exception in test/index.test.ts

  TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for ***/test/index.test.ts

  TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for ***/test/index.test.ts
      at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:160:9)
      at defaultGetFormat (node:internal/modules/esm/get_format:203:36)
      at defaultLoad (node:internal/modules/esm/load:143:22)
      at async ModuleLoader.load (node:internal/modules/esm/loader:403:7)
      at async ModuleLoader.moduleProvider (node:internal/modules/esm/loader:285:45)
      at async link (node:internal/modules/esm/module_job:76:21)

  ✘ test/index.test.ts exited with a non-zero exit code: 1
  ─

  1 uncaught exception
```

---

✅ And here's the file running successfully as `"commonjs"`:
```bash
npx ava
```

```
  ✘ [fail]: Number should not be a text
  ─

  Number should not be a text

  test/index.test.ts:2

   1: import test from "ava";                     
   2:                                             
   3: test("Number should not be a text", (t) => {

  Difference (- actual, + expected):

  - '1'
  + 1

  › test/index.test.ts:2:847

  ─

  1 test failed
```

**Comment by wellwelwel** [comment]
I found a similar issue (not necessarily related), but it looks was closed without a conclusion, despite the _PR_ #3192:

- #2593

**Comment by novemberborn** [comment]
```
  TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for ***/test/index.test.ts

  TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for ***/test/index.test.ts
      at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:160:9)
      at defaultGetFormat (node:internal/modules/esm/get_format:203:36)
      at defaultLoad (node:internal/modules/esm/load:143:22)
      at async ModuleLoader.load (node:internal/modules/esm/loader:403:7)
      at async ModuleLoader.moduleProvider (node:internal/modules/esm/loader:285:45)
      at async link (node:internal/modules/esm/module_job:76:21)
```

This stacktrace is internal to Node.js. My guess would be that the `tsx` loader isn't active. It may work for `commonjs` since AVA uses `require()` in that scenario, and maybe the loader does work there.

**Reply by wellwelwel** [reply]
Thanks, @novemberborn!

---

By removing `tsx` and keeping it as `"commonjs"`, it stop to work and returns a common error of a "non-compiled" **ts** file:

```
  Uncaught exception in test/index.test.ts

  ***/test/index.test.ts:1
  import test from 'ava';
  ^^^^^^

  SyntaxError: Cannot use import statement outside a module

  ✘ test/index.test.ts exited with a non-zero exit code: 1
  ─

  1 uncaught exception
```

---

Do you think there is a way to reach out it (**AVA** with **TypeScript** + **ESM** without compile) using other loader or a different approach?

Seeing the discussions in PR #2593, I saw that the recommendation was to use **tsx**, but I also tried **ts-node**, without success.

**Reply by novemberborn** [reply, marked as answer]
I intentionally avoid this mess myself, using `@ava/typescript` and always running the compiler in a split tab in my terminal.

**Reply by wellwelwel** [reply]
By using `@ava/typescript` (independent of using `tsx` or not), it causes this error:

```
  Uncaught exception in test/index.test.ts

  TypeError: Cannot destructure property 'negotiateProtocol' of 'undefined' as it is undefined.

  TypeError: Cannot destructure property 'negotiateProtocol' of 'undefined' as it is undefined.
      at typescriptProvider (***/node_modules/@ava/typescript/index.js:77:45)
      at run (***/node_modules/ava/lib/worker/base.js:227:11)
      at async ***/node_modules/ava/lib/worker/base.js:281:2

  ✘ test/index.test.ts exited with a non-zero exit code: 1
  ─

  1 uncaught exception
```

---

I tried using the recommendations from README.md:

```json
{
  "devDependencies": {
    "@ava/typescript": "^4.1.0",
    "ava": "^6.1.1"
  },
  "ava": {
    "files": [
      "test/**/*.test.ts"
    ],
    "require": [
      "@ava/typescript"
    ],
    "typescript": {
      "extensions": [
        "ts"
      ],
      "rewritePaths": {
        "test/": "build/"
      },
      "compile": false
    }
  }
}
```

---

@novemberborn, maybe should I open an Issue on https://github.com/avajs/typescript?

**Reply by wellwelwel** [reply]
For a better context, I was creating examples for the most popular Test Runners and how to reach out in the same result using them, including **AVA**:

```md
# Comparing Test Runners

**Comparison using **Node.js**, **TypeScript** (_no compile_) and **ESM** to show a simple error test:**

- Let's starting from installation 🔬
```

- Then, the examples.

**Reply by novemberborn** [reply]
It's misconfigured. Don't add the `require` block, and you don't need to specify `extensions` either.

**Reply by wellwelwel** [reply]
I get it now, `@ava/typescript` is compiling it before to run.

I'll keep the **AVA** example as `"commonjs"`, since the main idea is not to compile, and relate this discussion on it.

Thanks again, @novemberborn! 🚀

**Comment by wellwelwel** [comment]
> I intentionally avoid this mess myself, using `@ava/typescript` and always running the compiler in a split tab in my terminal.

Regardless, I will mark it as the answer, since this problem does not belong to this repository.

@novemberborn, sorry for the inconvenience (and also, thanks!) 🙋🏻‍♂️

**Comment by tommy-mitchell** [comment]
I use [`tsimp`](https://github.com/tapjs/tsimp) with this exact configuration (from the [recipes](https://github.com/avajs/ava/blob/main/docs/recipes/typescript.md)):

```json
"ava": {
	"extensions": {
		"ts": "module"
	},
	"nodeArguments": [
		"--import=tsimp"
	]
}
```

**Reply by wellwelwel** [reply]
Thanks @tommy-mitchell 🩵

In fact now it works for both `"commonjs"` and `"module"` skipping extra configs 🧑🏻‍🎓✨

---

**Note**

- In a first moment, [tsimp](https://github.com/tapjs/tsimp) is compiling **TypeScript** to `.tsimp` dir even including `noEmit` to `true`, then I just followed the [tsimp documentation](https://github.com/tapjs/tsimp?tab=readme-ov-file#config-file-changes-and-extends-options):

> Also, the following fields are always hard-coded by tsimp:
> 
> - `outDir` Because tsimp isn't a build tool, but rather a module
>   importer, it doesn't actually write the emitted JavaScript to
>   disk. (Ok, technically it does, but only as a cache.) So, the
>   `outDir` is hard-coded to `.tsimp-compiled`, but this is never
>   used.
> - [...]
> - `noEmit` is always set `false`, because the entire point is to
>   get the JavaScript code for Node to run. That said, the "emit"
>   is fully virtual, and nothing is written to disk (except to
>   avoid compiling the same code multiple times).

**Reply by iSpeakNerd** [reply]
@wellwelwel maybe a good idea to mark this tsimp answer as the solution? I'm currently digging around the internet for a ts-native test tool that can handle ESM as well. Jest is a freaking mess of configs, and no other solutions seem as simple to set up as what [I saw for Ava](https://github.com/avajs/ava/blob/main/docs/recipes/typescript.md)

**Reply by wellwelwel** [reply]
@iSpeakNerd, as it's an architectural decision, I've marked the **AVA** maintainer's own answer. Although this answer works perfectly, the focus of the discussion was on not compiling, but **tsimp** does it on disk (as a cache — not temporary).


**[OFF]**

> I'm currently digging around the internet for a ts-native test tool that can handle ESM as well.

When I opened this discussion, my intention was to create the same test case with 5 different test runners, including **AVA**. Maybe it helps you indirectly with your problem: [poku.io/docs/comparing#typescript-comparison](https://poku.io/docs/comparing#typescript-comparison).
